File: app/src/main/java/net/emilla/util/Features.java
package net.emilla.util;

import android.content.pm.PackageManager;
import android.os.Build;

public enum Features {
    ;

    public static boolean camera(PackageManager pm) {
        return pm.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY);
    }

    public static boolean phone(PackageManager pm) {
        return pm.hasSystemFeature(PackageManager.FEATURE_TELEPHONY);
    }

    public static boolean sms(PackageManager pm) {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU
            ? pm.hasSystemFeature(PackageManager.FEATURE_TELEPHONY_MESSAGING)
            : pm.hasSystemFeature(PackageManager.FEATURE_TELEPHONY);
    }

    public static boolean torch(PackageManager pm) {
        return pm.hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH);
        // todo: test this on a minSdk device.
    }

}

File: app/src/main/java/net/emilla/util/Features.kt
@file:JvmName("Features")

package net.emilla.util

import android.content.pm.PackageManager
import android.os.Build

@JvmName("camera")
fun PackageManager.hasCameraFeature(): Boolean {
    return hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)
}

@JvmName("phone")
fun PackageManager.hasPhoneFeature(): Boolean {
    return hasSystemFeature(PackageManager.FEATURE_TELEPHONY)
}

@JvmName("sms")
fun PackageManager.hasSmsFeature(): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
        hasSystemFeature(PackageManager.FEATURE_TELEPHONY_MESSAGING)
    } else {
        hasSystemFeature(PackageManager.FEATURE_TELEPHONY)
    }
}

@JvmName("torch")
fun hasTorchFeature(pm: PackageManager): Boolean {
    // todo: it'd be a good idea to test this on a minSdk device.
    return pm.hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH)
}
